### PR TITLE
Transaction profiles

### PR DESCRIPTION
**Pull request by tilk**
This PR adds transaction profiles. A profile contains a per-cycle information about:

* Which transactions/methods were running. For running methods, which transaction/method called.
* Which transactions wanted to run but couldn't, and which conflicting transaction was running instead.
* Which methods were locked, but not running.

TODO:

* [x] Calculating statistics for the entire profile.
* [x] Options, filtering and sorting in the command line tool.
* [x] Running profile generation in CI to check for regressions.
* [x] Non-inverted call graphs (transactions to called methods).
* [x] User documentation (document the script and the profile option)

In the future:

* Some way of visualizing profiles.

### CODE REVIEW

**Review by lekcyjna123 — COMMENTED** [review]
Please check if #552 can be useful.

**Comment by lekcyjna123** [comment]
According to the point:
> Some way of visualizing profiles.

I am currently thinking about two possible way:
- Something similar to waveform, where we have a one timeline for each transaction. It is green if it is running, blue when not ready and red if conflicting, ready and not running
- Graph between transactions and methods. Edges in such graph can visualise how often conflict on these methods/transactions caused a stall of one of them.

File: test/transactions/test_methods.py [hunk 1]
```diff
@@ -609,8 +608,10 @@ def process():
                 self.assertTrue(in1 != self.bad_number or not out_t1)
                 self.assertTrue(in2 != self.bad_number or not out_t2)
 
+                yield
+
         with self.run_simulation(self.circ, 100) as sim:
-            sim.add_process(process)
+            sim.add_sync_process(process)
```
[the remark this answers is not in the document]
Comment (tilk) [reply test/transactions/test_methods.py:614]: What exactly did it test before which isn't tested now? The change basically replaced real-time delay with a single-cycle delay, which shouldn't matter at all for combinational circuits. What did I miss? As written before, that test couldn't generate a profile.

[the remark this answers is not in the document]
Comment (lekcyjna123) [reply test/transactions/test_methods.py:614]: Before it tested multiple selection turns in one clock cycle (so if the input data will change in the same cycle, then output also should change). Now we have only one selection for one cycle, but you are right and it shouldn't matter, especially when it wasn't visible in profiling.

**Comment by tilk** [comment]
> I am currently thinking about two possible way:
> 
>     * Something similar to waveform, where we have a one timeline for each transaction. It is green if it is running, blue when not ready and red if conflicting, ready and not running
> 
>     * Graph between transactions and methods. Edges in such graph can visualise how often conflict on these methods/transactions caused a stall of one of them.

Both ideas are great, but I think this should be left for a future PR. For now, I'm writing a command-line script which will show transaction and method statistics using inverted call trees (methods to transactions). Normal call trees (transactions to methods) might be useful too.

File: transactron/profiler.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    def encode(self, file_name: str):
+        with open(file_name, "w") as fp:
+            fp.write(self.to_json())  # type: ignore
+
+    @staticmethod
+    def decode(file_name: str):
+        with open(file_name, "r") as fp:
+            return Profile.from_json(fp.read())  # type: ignore
+
+    def analyze_transactions(self: "Profile") -> list[TransactionStat]:
```
[the remark this answers is not in the document]
Comment (tilk) [reply transactron/profiler.py]: Removed the annotation, this was not necessary.

File: transactron/profiler.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    transactions_and_methods: dict[int, ProfileInfo] = field(default_factory=dict)
+    waiting_transactions: dict[int, dict[int, int]] = field(default_factory=dict)
+    running: dict[int, dict[int, Optional[int]]] = field(default_factory=dict)
+
+    def encode(self, file_name: str):
+        with open(file_name, "w") as fp:
+            fp.write(self.to_json())  # type: ignore
+
+    @staticmethod
+    def decode(file_name: str):
```
[the remark this answers is not in the document]
Comment (tilk) [reply transactron/profiler.py:126]: Added.

File: transactron/profiler.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    """Transactron execution profile.
+
+    Can be saved by the simulator, and then restored by an analysis tool.
+    In the profile data structure, methods and transactions are referred to
+    by their unique ID numbers.
+
+    Attributes
+    ----------
+    transactions_and_methods : dict[int, ProfileInfo]
+        Information about transactions and methods.
```
Comment (lekcyjna123) [inline transactron/profiler.py:114]:
```suggestion
        Mapping from transaction id or method id to the information about that transaction/method.
```

File: test/common/profiler.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        for method in method_map.methods:
+            profile.transactions_and_methods[get_id(method)] = ProfileInfo(
+                method.owned_name, local_src_loc(method.src_loc), False
+            )
+
+        cycle = 0
+
+        yield Passive()
+        while True:
+            yield Delay(5e-7)  # shorter than one clock cycle
```
Comment (lekcyjna123) [inline test/common/profiler.py:44]: Please, take the cycle length as an argument to `profiler_process`

**Review by lekcyjna123 — APPROVED** [review]

File: test/common/profiler.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            profile.transactions_and_methods[get_id(transaction)] = ProfileInfo(
+                transaction.owned_name, local_src_loc(transaction.src_loc), True
+            )
+
+        for method in method_map.methods:
+            profile.transactions_and_methods[get_id(method)] = ProfileInfo(
+                method.owned_name, local_src_loc(method.src_loc), False
+            )
+
+        cycle = 0
```
Comment (makz00) [inline test/common/profiler.py:40]: I cannot see the usage of `cycle` variable. Is it needed?

Comment (tilk) [reply test/common/profiler.py]: It's not, thank you for catching that.

**Review by makz00 — APPROVED** [review]

**Pull request merged**